File: 第三阶段/3.2-访问修饰符/access_modifiers.py
# ============================================================
# Python 访问修饰符 —— 对比 C# 理解 Python 的访问控制
# 对应文章：3.2 访问修饰符
# ============================================================
# 核心区别：C# 有语法级别的访问控制（public/private/protected/internal），
# Python 没有真正的访问控制，完全靠命名约定。
# ============================================================

# ============================================================
# 1. Python 的"访问修饰符"—— 纯靠约定
# ============================================================
# C# 等价：public / private / protected / internal
# Python 实际只有三个约定：
#   - 无前缀：public（公开）
#   - _ 单下划线：protected（约定私有，外部仍可访问）
#   - __ 双下划线：name mangling（名称改写，非真正私有）

class Person:
    def __init__(self, name, age):
        self.name = name          # public —— 任何地方都能访问
        self._age = age           # _ 前缀 —— 约定为"内部使用"
        self.__secret = "hidden"  # __ 前缀 —— 名称改写（name mangling）

    def greet(self):
        """public 方法 —— C# 等价：public void Greet()"""
        return f"Hello, I'm {self.name}"

    def _internal_helper(self):
        """_ 前缀方法 —— C# 等价：internal / protected"""
        return f"内部调用: age={self._age}"

    def __private_method(self):
        """__ 前缀方法 —— C# 等价：private"""
        return self.__secret


print("=" * 50)
print("1. Python 的访问约定")
print("=" * 50)

p = Person("Alice", 25)

# public —— 完全正常访问
print(f"public:  p.name = {p.name}")

# _ protected —— 能访问，但"约定不该访问"
print(f"_约定:   p._age = {p._age}")

# __ name mangling —— 直接访问会报错
try:
    _ = p.__secret
except AttributeError as e:
    print(f"__:      p.__secret -> {e}")

# 但通过名称改写后的名字仍可访问（Python 不做真正保护）
print(f"改写后:  p._Person__secret = {p._Person__secret}")

# 方法同理
print(f"public:  p.greet() = {p.greet()}")
print(f"_约定:   p._internal_helper() = {p._internal_helper()}")

try:
    _ = p.__private_method()
except AttributeError as e:
    print(f"__方法:  p.__private_method() -> {e}")

print(f"改写后:  p._Person__private_method() = {p._Person__private_method()}")


# ============================================================
# 2. 名称改写（Name Mangling）的目的
# ============================================================
# 名称改写不是为了安全性，而是为了避免子类意外覆盖父类成员
# C# 中用 private 实现相同目的——子类天然看不到父类的 private 成员

class Parent:
    def __init__(self):
        self.__secret = "parent's secret"  # 会变成 _Parent__secret

    def reveal(self):
        print(f"  Parent.reveal() -> {self.__secret}")


class Child(Parent):
    def __init__(self):
        super().__init__()
        self.__secret = "child's secret"   # 会变成 _Child__secret（不同名称！）

    def reveal(self):
        print(f"  Child.reveal()  -> {self.__secret}")


print("\n" + "=" * 50)
print("2. 名称改写的目的 —— 避免子类意外覆盖")
print("=" * 50)

parent = Parent()
child = Child()

print("parent.reveal():")
parent.reveal()
print("child.reveal():")
child.reveal()

# 两者互不干扰——这就是名称改写的价值
print(f"parent._Parent__secret = {parent._Parent__secret}")
print(f"child._Child__secret   = {child._Child__secret}")


# ============================================================
# 3. @property —— Python 实现"受控访问"的核心手段
# ============================================================
# C# 等价：private set + validation in setter
# Python 没有 private set，但 @property 的 setter 可以加验证逻辑

class BankAccount:
    def __init__(self, owner, balance=0):
        self.owner = owner
        self.__balance = balance  # "私有"余额

    @property
    def balance(self):
        """只读属性 —— C# 等价：public decimal Balance { get; }"""
        return self.__balance

    @balance.setter
    def balance(self, value):
        """带验证的 setter —— C# 等价：private set + 验证逻辑"""
        if value < 0:
            raise ValueError("余额不能为负数")
        self.__balance = value

    def deposit(self, amount):
        if amount <= 0:
            raise ValueError("存款金额必须为正")
        self.__balance += amount
        return f"存入 {amount}，余额 {self.__balance}"

    def withdraw(self, amount):
        if amount <= 0:
            raise ValueError("取款金额必须为正")
        if amount > self.__balance:
            raise ValueError("余额不足")
        self.__balance -= amount
        return f"取出 {amount}，余额 {self.__balance}"


print("\n" + "=" * 50)
print("3. @property —— Python 的'受控访问'")
print("=" * 50)

account = BankAccount("Alice", 1000)
print(f"户主: {account.owner}")
print(f"余额: {account.balance}")
print(account.deposit(500))
print(account.withdraw(200))
print(f"最终余额: {account.balance}")

# 尝试设置非法值
try:
    account.balance = -100
except ValueError as e:
    print(f"非法设置余额: {e}")

# 尝试直接修改"私有"余额——不推荐，但技术上可行
account._BankAccount__balance = 99999
print(f"直接修改后余额: {account.balance}  (不推荐！)")


# ============================================================
# 4. __all__ —— 模块级别的访问控制
# ============================================================
# C# 等价：internal / [InternalsVisibleTo]
# __all__ 控制 from module import * 时导出哪些名称
# 这是 Python 中唯一真正的"模块级"访问控制

# 假设这是一个模块的 __all__ 声明：
_all_example_names = ["PublicClass", "public_function"]
__private_module_var = "这是模块私有变量"

class PublicClass:
    """会被 __all__ 导出"""
    pass

class _PrivateClass:
    """约定私有——__all__ 不会包含它"""
    pass

def public_function():
    """公开函数"""
    return "我是公开函数"

def _private_function():
    """私有函数"""
    return "我是私有函数"


print("\n" + "=" * 50)
print("4. __all__ —— 模块级访问控制")
print("=" * 50)

# 演示 __all__ 的效果
print(f"__all__ 声明导出: {_all_example_names}")
print(f"public_function() = {public_function()}")
print(f"PublicClass = {PublicClass}")
print(f"_PrivateClass = {_PrivateClass}")          # 仍可访问，只是约定不访问
print(f"_private_function() = {_private_function()}")  # 同上

# 实际使用：
# from this_module import *  -> 只会导入 PublicClass 和 public_function
# import this_module         -> 所有名称都可通过 this_module.xxx 访问


# ============================================================
# 5. 名称约定汇总
# ============================================================

print("\n" + "=" * 50)
print("5. C# vs Python 访问控制对比总结")
print("=" * 50)

summary = """
| 概念           | C#           | Python                 |
|----------------|--------------|------------------------|
| 公开           | public       | 无前缀（name）         |
| 私有           | private      | __ 前缀（名称改写）    |
| 受保护         | protected    | _ 前缀（仅约定）       |
| 程序集内部     | internal     | __all__ 控制 * 导入    |
| 只读属性       | { get; }     | @property 只有 getter  |
| 带验证的属性   | private set  | @property + setter     |
| 常量           | const        | 全大写 + _ 分隔        |
| 枚举           | enum         | class + 类变量 或 Enum |
| 接口约束       | interface    | Protocol（typing）     |
"""
print(summary)

print("要点：Python 的访问控制全靠约定（'we are all consenting adults'）")
print("      C# 由编译器强制执行，Python 由程序员自觉遵守")

File: 第三阶段/3.2-访问修饰符/AccessModifiers.cs
// ============================================================
// C# 访问修饰符 —— 对比 Python 理解 C# 的访问控制
// 对应文章：3.2 访问修饰符
// ============================================================
// 核心区别：C# 有语法级别的访问控制（编译器强制执行），
// Python 完全靠命名约定（"we are all consenting adults"）。
// ============================================================

using System;

// ============================================================
// 1. 基本访问修饰符
// ============================================================
// Python 等价：无前缀 = public, _ = protected(约定), __ = name mangling

public class Person
{
    // public —— Python 等价：无前缀 self.name
    public string Name { get; set; }

    // private —— Python 等价：__ 前缀（仅靠约定，不强制）
    private int _age;

    // protected —— Python 等价：_ 前缀（仅靠约定）
    protected string Secret;

    // internal —— Python 等价：__all__ 控制模块导入
    internal string InternalNote = "internal data";

    public Person(string name, int age)
    {
        Name = name;
        _age = age;
        Secret = "hidden";
    }

    // private 方法 —— Python 等价：__method()
    private int GetAge()
    {
        return _age;
    }

    // protected 方法 —— Python 等价：_method()
    protected void RevealSecret()
    {
        Console.WriteLine($"  Secret: {Secret}");
    }

    // public 方法调用 private 和 protected
    public void ShowInfo()
    {
        Console.WriteLine($"  Name={Name}, Age={GetAge()}");  // 可以调用 private
        RevealSecret();  // 可以调用 protected
        Console.WriteLine($"  Internal: {InternalNote}");
    }
}

// ============================================================
// 2. 只读属性（C# 的受控访问）
// ============================================================
// Python 等价：@property 只有 getter

public class BankAccount
{
    public string Owner { get; }
    public decimal Balance { get; private set; }  // 只有类内部能 set

    public BankAccount(string owner, decimal initialBalance)
    {
        Owner = owner;
        Balance = initialBalance;
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("存款金额必须为正");
        Balance += amount;
        Console.WriteLine($"  存入 {amount}，余额 {Balance}");
    }

    public void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("取款金额必须为正");
        if (amount > Balance)
            throw new InvalidOperationException("余额不足");
        Balance -= amount;
        Console.WriteLine($"  取出 {amount}，余额 {Balance}");
    }
}

// ============================================================
// 3. sealed class —— 不可继承
// ============================================================
// Python 没有等价机制（Python 不限制继承）
// C# 用 sealed 阻止继承，常用于安全/性能优化

public sealed class FinalClass
{
    public string Message { get; } = "我是 sealed 类，不能被继承";

    public override string ToString() => $"FinalClass({Message})";
}

// sealed class 用于子类——标记不允许再被继承
public class Animal
{
    public virtual string Speak() => "...";
}

public sealed class Dog : Animal   // sealed：不能再有子类继承 Dog
{
    public override string Speak() => "汪汪！";
}

// class Puppy : Dog { }  // 编译错误！Dog 是 sealed

// ============================================================
// 4. abstract class —— 不能实例化，必须被继承
// ============================================================
// Python 等价：from abc import ABC, abstractmethod

public abstract class Shape
{
    public abstract double Area { get; }          // 抽象属性
    public abstract double CalculateArea();       // 抽象方法

    // 非抽象方法——可以有实现
    public void Describe()
    {
        Console.WriteLine($"  面积是 {CalculateArea():F2}");
    }
}

public class Circle : Shape
{
    public double Radius { get; }

    public Circle(double radius) => Radius = radius;

    public override double Area => CalculateArea();

    public override double CalculateArea()
        => Math.PI * Radius * Radius;
}

// ============================================================
// 5. Main 方法 —— 运行演示
// ============================================================

public static class Program
{
    public static void Main()
    {
        // 1. 基本访问修饰符
        Console.WriteLine("=" + new string('=', 49));
        Console.WriteLine("1. C# 四种访问修饰符");
        Console.WriteLine("=" + new string('=', 49));

        var p = new Person("Alice", 25);
        Console.WriteLine($"public:  p.Name = {p.Name}");
        // Console.WriteLine(p._age);      // 编译错误！private
        // Console.WriteLine(p.Secret);     // 编译错误！protected
        Console.WriteLine($"internal: p.InternalNote = {p.InternalNote}");

        Console.WriteLine("通过 public 方法访问所有成员:");
        p.ShowInfo();

        Console.WriteLine("\n对比 Python: C# 编译器强制检查，Python 运行时才报错");
        Console.WriteLine("  C#:  private x;   -> 直接访问 -> 编译错误");
        Console.WriteLine("  Py:  self.__x     -> 直接访问 -> AttributeError (运行时)");

        // 2. 只读 / private set
        Console.WriteLine();
        Console.WriteLine("=" + new string('=', 49));
        Console.WriteLine("2. private set —— Python @property setter 等价");
        Console.WriteLine("=" + new string('=', 49));

        var account = new BankAccount("Alice", 1000);
        Console.WriteLine($"户主: {account.Owner}");
        Console.WriteLine($"余额: {account.Balance}");
        account.Deposit(500);
        account.Withdraw(200);
        Console.WriteLine($"最终余额: {account.Balance}");
        // account.Balance = -100;  // 编译错误！private set
        // 对比 Python: 可以 account._BankAccount__balance = 99999（绕过保护）

        // 3. sealed class
        Console.WriteLine();
        Console.WriteLine("=" + new string('=', 49));
        Console.WriteLine("3. sealed class —— 不可继承（Python 无等价机制）");
        Console.WriteLine("=" + new string('=', 49));

        var final = new FinalClass();
        Console.WriteLine(final);
        Console.WriteLine("C# sealed vs Java final class vs Python: 无限制");

        var dog = new Dog();
        Console.WriteLine($"Dog.Speak() = {dog.Speak()}");
        Console.WriteLine("Dog 被 sealed，不能再有子类");

        // 4. abstract class
        Console.WriteLine();
        Console.WriteLine("=" + new string('=', 49));
        Console.WriteLine("4. abstract class —— Python 等价：ABC + @abstractmethod");
        Console.WriteLine("=" + new string('=', 49));

        // var s = new Shape();  // 编译错误！不能实例化抽象类
        var circle = new Circle(5);
        Console.WriteLine($"Circle(radius={circle.Radius})");
        circle.Describe();

        // 5. 对比总结
        Console.WriteLine();
        Console.WriteLine("=" + new string('=', 49));
        Console.WriteLine("5. Python vs C# 访问控制对比总结");
        Console.WriteLine("=" + new string('=', 49));
        Console.WriteLine(@"
| 概念           | C#            | Python                  |
|----------------|---------------|-------------------------|
| 公开           | public        | 无前缀                  |
| 私有           | private       | __ 前缀（约定）         |
| 受保护         | protected     | _ 前缀（约定）          |
| 程序集内部     | internal      | __all__ 控制模块导入    |
| 强制执行       | 编译器检查    | 运行时才会报错          |
| 只读属性       | { get; }      | @property 只有 getter   |
| 带验证的 set   | private set   | @property setter        |
| 不可继承       | sealed class  | 无（Python 不限制继承） |
| 抽象类         | abstract      | ABC + @abstractmethod   |
");
    }
}
